### ISSUE TITLE
Evaluating custom model

### ISSUE THREAD

**Issue by jinseokbae** [AIGAnimation/CAMDM#8]
Hi,
thanks for releasing the codebase for the amazing work.

I slightly modified the model and want to get metric numbers with the newly trained model.
However, the current version of the evaluation code assumes fixed set of a number of experiments and I cannot find the way to evaluate my own model.
How can I prepare evaluation data from the custom model so that I can add my own experiment in the evaluation codes?

Best regards,

**Comment by jinseokbae** [comment]
Additionally, I'm curious if there is a way to measure the performance purely with the Python implementation. I expect there can be an inference code implemented with PyTorch.

**Comment by Shimingyi** [comment]
Hi @jinseokbae , thanks for the question. Our system consists of two modules: AR model training in PyTorch and the runtime part in Unity. It would not be fair if we simply output the motion using Python without all the runtime strategies. However, re-implementing all these features in Python is non-trivial.

For the evaluation, we suggest exporting the results using a Unity script. The script should include two functions: 1. Loading the predefined control signals and 2. Recording the animation to an exported BVH file. We can provide you with an example of this script by the end of this week, which you can then modify to suit your character structure.

**Comment by jinseokbae** [comment]
Hi @Shimingyi, thanks for the rapid reply.
It would be extremely excited to have the script to run the predefined control signal so that fairly compare the custom model with the released experiments.
Also, I hugely appreciate for the effort on the code work.
Looking forward to the script.

Best,

**Comment by Shimingyi** [comment]
Hi @jinseokbae , sorry for the late update. I have updated the unity code to support the 'Control signal loading' and 'Runtime motion exporting', you can check the details in the latest two commits! 

<img width="629" alt="Image" src="https://github.com/user-attachments/assets/0b6cf22d-58c8-4765-a65d-f51ed2c312d4" />

**Comment by jinseokbae** [comment]
Hi @Shimingyi,
I really appreciate for sharing this code.
Now I can successfully load my model and see the results that are controlled with the preset controls!

One more question here.
Currently, it seems that the user should manually switch the styles of motions in order to extract generated motions for 100 styles of motions.
However it is hard to make all the bvh file to have motion sequence in equal time length.
Is it okay to compare two models with different number of frames for generated motions in each style?

**Comment by Shimingyi** [comment]
Hi @jinseokbae , sorry I missed one line in the code; now it supports the transition to the new style after running out of all preset controls: https://github.com/AIGAnimation/CAMDM/blob/main/Unity/Assets/Demo/CAMDM/BioAnimation.cs#L341 

But I found the exporting function may have bugs on the rotation transformation, fixing it now.

**Comment by jinseokbae** [comment]
Hi @Shimingyi, thanks for the fast reply!

Also, I would really appreciate for the rest of the fix.
Looking forward to the updates!

[Updated]
I just recognized that python codes for evaluation (e.g. compute_feat.py) cannot load exported bvh files.
Errors come from `load_bvh()` function in `bvh_motion.py`.

I didn't change character, but just used default y-bot character in Unity.
It would be also appreciated if you can verify whether there is some additional step required to use these bvh files for evaluation.

This is one of the exported bvh files I've extracted from running Unity code.
[motion_eval_Aeroplane.zip](https://github.com/user-attachments/files/18946794/motion_eval_Aeroplane.zip)

As long as I found, the extracted bvh file is weird.
Although it contains long sequence of motions (I used given preset.txt), only 20 frames are counted in the BVH file (see Line 139 in the attached file).
Also, as character includes 23 joints (1 for root 22 for other joints), each line under the "Motion" section in the bvh file should be represented with 72 numbers (6 * 1 for root, 22 * 3 for other joints).
However extracted BVH file has 78 dimension in total.

I think BVH exporter should be also revised.

**Comment by Shimingyi** [comment]
We have fixed the bugs in the BVH exporting function. A wrong bvh_header causes it. To avoid potential confusion, now we generate the raw text of hierarchical information from the character directly. Here is my visualization in the Blender.

![Image](https://github.com/user-attachments/assets/df3de39f-3ac4-4e4e-a191-fabf6ad08904)

The error in the evaluation code might be because of the different joint names in your custom character; you need to add a new joint configuration according to your own model's joint name:

Refer to:
Existing joint configuration, you need to add new to L46, L57(remember to keep the same joint number of 15) https://github.com/AIGAnimation/CAMDM/blob/main/Evaluation/feature_extractor/process_train_data.py#L25
Entrance: 
https://github.com/AIGAnimation/CAMDM/blob/main/Evaluation/compute_feat.py#L22

Because this part is applicable to our provided results data, so we might not modify it to adapt our current 3D model in Unity.

**Comment by jinseokbae** [comment]
Thanks for the update.

I'm not using custom character model, but just using default character provided in this codebase.
I can visualize extracted bvh files in Blender, as attached in the message.

However, can you reproduce the numbers in the paper with the bvh files extracted from the updated code?
I'm trying it with the provided best model weights (not my own), but the resulting FID score seems very bad.

Let me clarify my current attempt:
1. I exported bvh files by running the released code (I didn't change any code in Unity, neither use custom characters. Also I just used pretrained weights provided by authors.)
2. I relocated bvh files under `Evaluation/data/exp1_test` folder and run `compute_feat.py`
    * I needed to add a few lines as the joint names in bvh files have different namings (e.g. "mixamorig:Hip"). I think this should be also updated in the master branch.
    * After this modification, npz files are created in the folder.
3. Then I executed `eval_motion_FID.py` with the new folder, and FID score is extremely bad (around 35).

**Comment by Shimingyi** [comment]
It might because the scale of new produced bvh is not same as previous version (current in meter but previous in centimetre), but I will check it tomorrow.

**Comment by Shimingyi** [comment]
I have made some modifications to the code, to ensure alignment between the newly exported BVH files and the evaluation samples we provided. While the speed looks a little bit different, you can get comparable FID based on your own exporting. 

![Image](https://github.com/user-attachments/assets/c99ebf2e-f202-41fc-bbc3-c853419f110e)
Yellow -> The new exporting; Grey -> old samples.

**Comment by jinseokbae** [comment]
Thanks for the update!

I'm still observing a suboptimal FID score (around 2) for the default model, which I retrained using the default options from the master branch.

Additionally, I discovered a potential issue in the evaluation code. In **eval_traj_alignment.py**, two metrics are computed to assess trajectory and orientation errors. Both metrics currently calculate the error as the difference between scalar angles projected on the ground, i.e., |θ_preset – θ_character|, by extracting angles using the **arctan2** function.

However, I've noticed that **arctan2** sometimes produces discontinuous values over time, leading to erroneously high error measurements—even when the character’s trajectory closely matches the preset trajectory. To obtain a more stable and reliable measurement, I suggest using the **arccos** function applied to the dot product of the corresponding 2D vectors, instead of directly subtracting scalar angles.

Looking forward to your thoughts on this approach.

**Comment by Shimingyi** [comment]
I can imagine the difficulties of reproducing the exact number because this system passed a few iterations after the paper submission: 

Before the open-sourcing, we always used raw 100style motion data to drive a slightly modified ybot; you can find it in our official video (the foot mesh is weird). But in the current version, to help others to drive another mixamo-type character, we decide to retarget the raw motion from 100style skeleton to mixamo-skeleton, and then we demo it with raw mixamo character. Besides the updates of model and character, we made a few modifications in our unity code, which also brings some variations. 

Most importantly, when we calculate the FID, we should train a feature extractor based on the GT dataset. In the default evaluation [codes](https://github.com/AIGAnimation/CAMDM/blob/main/Evaluation/feature_extractor/process_train_data.py#L80), you can see we only involve two datasets (raw 100style data + raw ybot data in MANN). However, in the current version, our retargeted mixamo skeleton does not appear in this GT distribution, and there is a high risk of getting a higher FID.

```
    data_folders = [
        ["./data/raw", "style_skeleton"],
        ["./data/raw_mann", "ybot_skeleton"]
    ]
```

When we decide to open-source all parts of this system, we can imagine such difficulties; that is the reason we back up the last version of our evaluation data and share it directly. 

Moreover, your suggestion for the metric calculation is right! I will modify our code to better align with the evaluation. Then, I plan to re-run the entire evaluation section and give an updated comparison table. I might not do it immediately, so you can use the results you produced first; just remember to train the classifier with the new batch of data. 

Lastly, before this repo, I don't think there was such a comprehensive framework to train/infer/evaluate the locomotion controller. AI4Animation is great, but their Unity-based data processing is so complex that it is not convenient to be a benchmark. We hope this repo can make it to be easier, and so welcome you to follow the entire practice, and share your advice to improve it : )

**Comment by jinseokbae** [comment]
That makes sense!
Thanks for the detailed explanation.

I'm really fascinated with all the efforts you've made in this repo.
And also, I appreciate for the rapid updates.
It was such a great experience for me to train/run the animation controller powered by the motion diffusion model.
I'm looking forward to your future work as well :)

I close this issue as the most of my questions has been resolved.
